scrape/scrape.go

```diff
@@ -999,7 +1005,23 @@ mainLoop:
 		default:
 		}
 
-		last = sl.scrapeAndReport(interval, timeout, last, errc)
+		// Temporary workaround for a jitter in go timers that causes disk space
+		// increase in TSDB.
+		// See https://github.com/prometheus/prometheus/issues/7846
+		scrapeTime := time.Now()
+		if interval > 100*scrapeTimestampTolerance {
+			// For some reason, a ticker might have been skipped, in which case we
+			// would call alignedScrapeTime.Add(interval) multiple times.
+			for scrapeTime.Sub(alignedScrapeTime) >= interval {
+				alignedScrapeTime = alignedScrapeTime.Add(interval)
+			}
+			// Align the scrape time if we are in the tolerance boundaries.
+			if t := scrapeTime.Sub(alignedScrapeTime); t >= -scrapeTimestampTolerance && t <= scrapeTimestampTolerance {
```

Comment:
The `t >= -scrapeTimestampTolerance` seems to be a noop, it'll always be true.

Reply:
I am protecting against the ticker going a bit upfront, like in my test:

$ go run test.go
1601037886576 / 1601037886576     
1601037887576 / 1601037887576     
1601037888576 / 1601037888576     
1601037889576 / 1601037889576     
1601037890576 / 1601037890576     
1601037891576 / 1601037891576     
1601037892576 / 1601037892576     
1601037893575 / 1601037893576     
1601037894576 / 1601037894576     
**1601037895575 / 1601037895575**
1601037896575 / 1601037896576     
**1601037897575 / 1601037897575**

Reply:
Do you mean going backwards?

That seems like something we should be handling independently of this, as we should never have two scrapes started within the same interval.

Reply:
Better idea: take alignedScrapeTime before the ticker is created